File: Exo5.5.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

int compteur = 0;      // compteur global
pthread_mutex_t mutex;  // mutex pour protéger compteur
int limite;             // limite passée en argument

void* incrementer(void* arg) 
{
    while (1) 
    {
        sleep(rand() % 5 + 1); // 1 à 5 secondes

        pthread_mutex_lock(&mutex);
        if (compteur >= limite) 
        {
            pthread_mutex_unlock(&mutex);
            break;
        }
        compteur++;
        pthread_mutex_unlock(&mutex);
    }
    return (NULL);
}

void* afficher(void* arg) 
{
    while (1)
    {
        sleep(2);
        pthread_mutex_lock(&mutex);
        if (compteur >= limite) 
        {
            pthread_mutex_unlock(&mutex);
            break;
        }
        printf("Compteur = %d\n", compteur);
        pthread_mutex_unlock(&mutex);
    }
    return (NULL);
}

int main(int argc, char* argv[])
{
    if (argc != 2)
    {
        printf("Usage: %s <limite>\n", argv[0]);
        return (1);
    }

    limite = atoi(argv[1]);
    srand(time(NULL));
    pthread_mutex_init(&mutex, NULL);

    pthread_t th_inc, th_aff;

    pthread_create(&th_inc, NULL, incrementer, NULL);
    pthread_create(&th_aff, NULL, afficher, NULL);

    pthread_join(th_inc, NULL);
    pthread_join(th_aff, NULL);

    printf("Compteur final = %d\n", compteur);
    pthread_mutex_destroy(&mutex);

    return (0);
}

File: Exo5.5.py
import threading
import random
import time
import sys

compteur = 0
limite = int(sys.argv[1]) if len(sys.argv) > 1 else 10
lock = threading.Lock()

def incrementer():
    global compteur
    while True:
        time.sleep(random.randint(1, 5))
        with lock:
            if compteur >= limite:
                break
            compteur += 1

def afficher():
    global compteur
    while True:
        time.sleep(2)
        with lock:
            if compteur >= limite:
                break
            print("Compteur =", compteur)

th_inc = threading.Thread(target=incrementer)
th_aff = threading.Thread(target=afficher)

th_inc.start()
th_aff.start()

th_inc.join()
th_aff.join()

print("Compteur final =", compteur)